src/internal/_equals.js

```diff
@@ -1,67 +1,95 @@
+var _arrayFromIterator = require('./_arrayFromIterator');
 var _has = require('./_has');
 var identical = require('../identical');
 var keys = require('../keys');
 var type = require('../type');
 
-// The algorithm used to handle cyclic structures is
-// inspired by underscore's isEqual
-module.exports = function _equals(a, b, stackA, stackB) {
-  var typeA = type(a);
-  if (typeA !== type(b)) {
-    return false;
-  }
-
-  if (typeA === 'Boolean' || typeA === 'Number' || typeA === 'String') {
-    return typeof a === 'object' ?
-      typeof b === 'object' && identical(a.valueOf(), b.valueOf()) :
-      identical(a, b);
-  }
 
+module.exports = function _equals(a, b, stackA, stackB) {
   if (identical(a, b)) {
     return true;
   }
 
-  if (typeA === 'RegExp') {
-    // RegExp equality algorithm: http://stackoverflow.com/a/10776635
-    return (a.source === b.source) &&
-           (a.global === b.global) &&
-           (a.ignoreCase === b.ignoreCase) &&
-           (a.multiline === b.multiline) &&
-           (a.sticky === b.sticky) &&
-           (a.unicode === b.unicode);
+  if (type(a) !== type(b)) {
+    return false;
   }
 
-  if (Object(a) === a) {
-    if (typeA === 'Date' && a.getTime() !== b.getTime()) {
+  switch (type(a)) {
+    case 'Null':
+    case 'Undefined':
       return false;
-    }
-
-    var keysA = keys(a);
-    if (keysA.length !== keys(b).length) {
+    case 'Array':
+    case 'Object':
+      break;
+    case 'Boolean':
+    case 'Number':
+    case 'String':
+      if (!(typeof a === typeof b && identical(a.valueOf(), b.valueOf()))) {
+        return false;
+      }
+      break;
+    case 'Date':
+      if (!identical(a.valueOf(), b.valueOf())) {
+        return false;
+      }
+      break;
+    case 'RegExp':
+      if (!(a.source === b.source &&
+            a.global === b.global &&
+            a.ignoreCase === b.ignoreCase &&
+            a.multiline === b.multiline &&
+            a.sticky === b.sticky &&
+            a.unicode === b.unicode)) {
+        return false;
+      }
+      break;
+    case 'Map':
+    case 'Set':
+      if (!_equals(_arrayFromIterator(a.entries()), _arrayFromIterator(b.entries()), stackA, stackB)) {
+        return false;
+      }
+      break;
+    case 'Int8Array':
+    case 'Uint8Array':
+    case 'Uint8ClampedArray':
+    case 'Int16Array':
+    case 'Uint16Array':
+    case 'Int32Array':
+    case 'Uint32Array':
+    case 'Float32Array':
+    case 'Float64Array':
+      break;
+    case 'ArrayBuffer':
+      break;
```

Comment:
Does it even make sense to apply the notion of value-based equality to objects of these types?